### PR TITLE
Fix FormattingHelpers.CountDigits(UInt128) and add more tests for Int128/UInt128.ToString

### LINKED ISSUE #74486: Crash/memory corruption when converting `Int128` to string
### Description

I found some bugs when trying to convert certain `Int128`/`UInt128` values to strings.

There are particular issues when the the value of the upper 64 bits is less than 5. If that is the case and the value of the lower 64 bits is greater than 9, the program usually crashes due to memory corruption. If the upper 64 bits are less than 5 and the lower 64 bits have a value less than 10, the upper bits are ignored and the resulting string only contains a single digit (when it should have ~20 digits).

### Reproduction Steps

1. For incorrect output:
```
Console.WriteLine(((UInt128) 1 << 64) + 3);
```

2. For buffer overflow/crash (any of the following lines):
```
Console.WriteLine(new UInt128(1, 10));
Console.WriteLine(new Int128(1, 10));
Console.WriteLine(new UInt128(5, 1));
Console.WriteLine(new Int128(5, 1));
```

### Expected behavior

1. Decimal value of 2<sup>64</sup> + 3 (i.e., `18446744073709551619`) written to console
2. Large numeric value(s) written to console

### Actual behavior

1. Console output: `3`

2. Exception:
```
An unhandled exception of type 'System.AccessViolationException' occurred in System.Private.CoreLib.dll: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'
   at System.Buffer.Memmove(Byte& dest, Byte& src, UIntPtr len)
   at System.IO.StreamWriter.WriteLine(String value)
   at System.IO.TextWriter.SyncTextWriter.WriteLine(Object value)
   at System.Console.WriteLine(Object value)
   at Program.<Main>$(String[] args)
```

### Regression?

Looks like this has been present since initial support for `Int128`: https://github.com/dotnet/runtime/commit/7f9ed8f271df8fb83862b3812d9c490e4ff74711

### Configuration

.NET 7.0 preview 7 (7.0.100-preview.7.22377.5) on Windows 11 Enterprise build 22622.575 (x64)

As of posting this, the issue seems to still be present in [`main`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs#L17-L20)

### Other information

It looks like the `FormattingHelpers.CountDigits(UInt128)` method incorrectly calculates the number of digits needed for string representations of 128-bit integer types. In certain cases it underestimates the necessary buffer length, leading to memory corruption when data is written past the end of the buffer later on.

https://github.com/dotnet/runtime/blob/e714888c563802ad372b8c3f703d99df46d70786/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs#L13-L36

`Int128` has two properties, `Upper` and `Lower`, which provide access to the high and low 64-bit halves of the 128-bit value respectively. `CountDigits` does not take `Upper` into account when its value is less than 5 and instead returns only the number of digits needed for `Lower`. This return value is directly used by `UInt128ToDecStr(UInt128)` as the length of the buffer it allocates for the output string (unless the length is 1, in which case it returns a constant single-digit string).

https://github.com/dotnet/runtime/blob/e714888c563802ad372b8c3f703d99df46d70786/src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs#L2220-L2239

There are also problems when `Upper` equals 5. The decimal representation of `5 << 64` has 20 digits, not 19. (`CountDigits` could probably use some comments&hellip;)

I actually first encountered this problem in PowerShell 7.3.0-preview.7, with crashes and corrupted output resulting from something along the lines of `([Int128] 1 -shl 66) + 10`

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves #74486

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs [hunk 1]
```diff
@@ -14,22 +14,37 @@ public static int CountDigits(UInt128 value)
...
+            // We have more than 1e19, so we have at least 20 digits
+            int digits = 20;
 
             if (upper > 5)
             {
-                digits++;
+                // ((2^128) - 1) / 1e20 < 34_02_823_669_209_384_635 which
+                // is 18.5318 digits, meaning the result definitely fits
+                // into 64-bits and we only need to add the lower digit count
+
```
Comment (dakersnar) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs:35]:
```suggestion
                // Here we divide the value by 1e20 and count the remaining digits from the result, 
                // all of which should fit within value.Lower, as shown:
                // ((2^128) - 1) / 1e20 is less than 34_02_823_669_209_384_635, 
                // which has 18.5318 digits, meaning the result of the division 
                //definitely fits into 64-bits for any value.

```
Adjusted for clarity, feel free to tweak further.

**Review by dakersnar — APPROVED** [review]
LGTM, slight tweak to comments

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs (hunk 1 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs:35]: I'll submit a separate PR to cleanup the comment more.

**Pull request merged**